npm/podController.go

```diff
@@ -146,162 +155,356 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
+	<-stopCh
+	log.Logf("Shutting down Pod workers")
+
+	return nil
+}
+
+func (c *podController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *podController) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// Pod resource to be synced.
+		// TODO : may consider using "c.queue.AddAfter(key, *requeueAfter)" according to error type later
+		if err := c.syncPod(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		log.Logf("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncPod compares the actual state with the desired, and attempts to converge the two.
+func (c *podController) syncPod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
 		return nil
 	}
 
-	npmPodObj, podErr := newNpmPod(podObj)
-	if podErr != nil {
-		metrics.SendErrorLogAndMetric(util.PodID, "[AddPod] Error: failed to create namespace %s, %+v with err %v", podObj.ObjectMeta.Name, podObj, podErr)
-		return podErr
+	// Get the Pod resource with this namespace/name
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	// lock to complete events
+	// TODO: Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("pod '%s' in work queue no longer exists", key))
+			// find the pod object from a local cache and start cleaning up process (calling cleanUpDeletedPod function)
+			cachedPod, found := c.npMgr.PodMap[key]
+			if found {
+				// TODO: better to use cachedPod when calling cleanUpDeletedPod?
+				err = c.cleanUpDeletedPod(getPodObjFromNpmObj(cachedPod))
+				if err != nil {
+					// cleaning process was failed, need to requeue and retry later.
+					return fmt.Errorf("Cannot delete ipset due to %s\n", err.Error())
+				}
+			}
+		}
+		return err
 	}
 
+	err = c.syncAddAndUpdatePod(pod)
```

Comment:
before this we should check for deletetimestamp and clean up



	if pod.DeletionTimestamp != nil ||
		pod.DeletionGracePeriodSeconds != nil {
		err := c.cleanUpDeletedPod(pod)
return err

	}

Reply:
applied